apps/pipelines/nodes/nodes.py

```diff
@@ -51,32 +60,68 @@ def _process(self, state: PipelineState) -> PipelineState:
         output = llm.invoke(state["messages"][-1], config=self._config)
         return output.content
 
-    def get_chat_model(self):
+    def get_llm_service(self):
         from apps.service_providers.models import LlmProvider
 
-        provider = LlmProvider.objects.get(id=self.llm_provider_id)
         try:
-            service = provider.get_llm_service()
-            return service.get_chat_model(self.llm_model, self.llm_temperature)
+            provider = LlmProvider.objects.get(id=self.llm_provider_id)
+            return provider.get_llm_service()
         except LlmProvider.DoesNotExist:
             raise PipelineNodeBuildError(f"LLM provider with id {self.llm_provider_id} does not exist")
         except ServiceProviderConfigError as e:
             raise PipelineNodeBuildError("There was an issue configuring the LLM service provider") from e
 
+    def get_chat_model(self):
+        return self.get_llm_service().get_chat_model(self.llm_model, self.llm_temperature)
+
 
-class CreateReport(LLMResponse):
-    __human_name__ = "Create a report"
+class LLMResponseWithPrompt(LLMResponse):
+    __human_name__ = "LLM response with prompt"
 
+    source_material_id: SourceMaterialId | None = None
     prompt: str = (
         "Make a summary of the following text: {input}. "
         "Output it as JSON with a single key called 'summary' with the summary."
     )
 
     def _process(self, state: PipelineState) -> PipelineState:
-        chain = PromptTemplate.from_template(template=self.prompt) | super().get_chat_model()
-        output = chain.invoke(state["messages"][-1], config=self._config)
+        prompt = PromptTemplate.from_template(template=self.prompt)
+        context = self._get_context(state, prompt)
+        chain = prompt | super().get_chat_model()
+        output = chain.invoke(context, config=self._config)
         return output.content
 
+    def _get_context(self, state: PipelineState, prompt: PromptTemplate):
```

Comment:
I'm wondering if this context shouldn't come from the [state](https://github.com/dimagi/open-chat-studio/blob/b88c5797dad31fb945d34a7661287b6e92331366/apps/service_providers/llm_service/state.py#L83-L98) that we use for the normal runnables in order to try and minimize duplication when trying to support experiment and pipeline based bots. It would have to be changed a bit, since the state expects the experiment as input as well.

Reply:
I think it's fine for now, if it was exact duplicate I'd say yes but it would be a bit different since the source material is linked to the node. But I guess you could extract a utility function that's used in both places:

```python
def get_prompt_context(session, prompt, source_material_id):
```

Reply:
Yeah, I wasn't sure how best to do this - that state is pretty well coupled to how the TopicBot uses it, and all the inheritance in those state classes was becoming a bit confusing to me.  We can discuss quickly the best way of refactoring / implementing this tomorrow.